### PR TITLE
[DB-30189] Fix GoLang E2E test

### PR DESCRIPTION
**Pull request by thebithead**
The changes here fix the GoLang based e2e test.  This test cases now works in the following scenarios:

- On a full K8s or OpenShift Cluster.
- On minikube (both locally and on Travis-CI) by passing the "-short" test option.  The "-short" test option skips the tests for Insights Server.
- There is an option to run the operator locally and run a single test within GoLand IDE. Example program arguments: 

-test.run TestNuodbAdmin -singleNamespace -test.v -localOperator -kubeconfig /home/tgates/test-kubeconfig -root /home/tgates/gopath/src/github.com/nuodb/nuodb-operator -globalMan /home/tgates/gopath/src/github.com/nuodb/nuodb-operator/deploy/operator-dev.yaml -namespacedMan /home/tgates/gopath/src/github.com/nuodb/nuodb-operator/deploy/role_binding.yaml 

The changes here are the minimum necessary to get the test working.  I made some minor improvements, but Improving the tests is NOT the goal DB-30189.

### CODE REVIEW

File: test/e2e/nuodb_database_base_test.go [hunk 1]
```diff
@@ -22,22 +22,22 @@ var (
...
-	smMemory           = "500m"
+	smMemory           = "500Mi"
 	smCount           int32 = 1
 	smCpu              = "100m"
 	smStorageSize           = "20G"
 	smStorageClass          = "local-disk"
 	engineOptions           = ""
 	teCount           int32 = 1
-	teMemory           = "100m"
+	teMemory           = "100Mi"
```
Comment (vegichan) [inline test/e2e/nuodb_database_base_test.go]: 100MBi might be a bit low. But I am guessing that you are not actually setting the real memory size for the TE?

File: test/e2e/nuodb_database_base_test.go [hunk 2]
```diff
@@ -61,12 +61,32 @@ func TestNuodbDatabase(t *testing.T) {
...
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	f := framework.Global
+
+	err = f.Client.Get(goctx.TODO(), types.NamespacedName{Name: "nuoadmin", Namespace: namespace}, exampleNuodbAdmin)
+	if err != nil {
+		t.Fatal(err)
+	}
```
Comment (vegichan) [inline test/e2e/nuodb_database_base_test.go]: assertNilError

File: test/e2e/nuodb_database_base_test.go [hunk 3]
```diff
@@ -61,12 +61,32 @@ func TestNuodbDatabase(t *testing.T) {
...
+		ApiServer:         apiServer,
+		Container:         container,
+	}
+
+	exampleNuodbAdmin := testutil.NewNuodbAdmin(namespace, adminSpec)
+	testutil.SetupOperator(t,ctx)
+	err = testutil.DeployNuodbAdmin(t, ctx, exampleNuodbAdmin )
+	if err != nil {
+		t.Fatal(err)
+	}
```
Comment (vegichan) [inline test/e2e/nuodb_database_base_test.go]: assertNilError

File: test/e2e/nuodb_database_base_test.go [hunk 4]
```diff
@@ -79,20 +99,17 @@ func TestNuodbDatabase(t *testing.T) {
...
 		t.Fatal(err)
 	}
 
-	f := framework.Global
 
-	err = f.Client.Get(context.TODO(), types.NamespacedName{Name: "nuodb", Namespace: namespace}, exampleNuodb)
+	err = f.Client.Get(goctx.TODO(), types.NamespacedName{Name: "nuodb", Namespace: namespace}, exampleNuodb)
 	if err != nil {
 		t.Fatal(err)
 	}
```
Comment (vegichan) [inline test/e2e/nuodb_database_base_test.go]: assertNilError

File: test/e2e/nuodb_insights_test.go [hunk 5]
```diff
@@ -166,12 +177,17 @@ func TestNuodbInsights(t *testing.T) {
...
 		Container:         container,
 	}
 
-	exampleNuodb := testutil.NewNuodbCluster(namespace, clusterSpec)
+	exampleNuodbAdmin := testutil.NewNuodbAdmin(namespace, nuodbAdminSpec)
 	testutil.SetupOperator(t,ctx)
+	err = testutil.DeployNuodbAdmin(t, ctx, exampleNuodbAdmin )
+	if err != nil {
+		t.Fatal(err)
+	}
```
Comment (vegichan) [inline test/e2e/nuodb_insights_test.go]: assertNilError

File: test/e2e/nuodb_insights_test.go [hunk 6]
```diff
@@ -166,12 +177,17 @@ func TestNuodbInsights(t *testing.T) {
...
+
+	exampleNuodb := testutil.NewNuodbDatabase(namespace, nuodbSpec)
 	err = testutil.DeployNuodb(t, ctx, exampleNuodb )
 	if err != nil {
 		t.Fatal(err)
```
Comment (vegichan) [inline test/e2e/nuodb_insights_test.go]: assertNilError

File: test/e2e/nuodb_scaleup_test.go [hunk 7]
```diff
@@ -24,14 +24,14 @@ func TestNuodbScale(t *testing.T) {
...
-		smMemory           = "500m"
+		smMemory           = "500Mi"
 		smCount           int32 = 1
 		smCpu              = "100m"
 		smStorageSize           = "20G"
 		smStorageClass          = "local-disk"
 		engineOptions           = ""
 		teCount           int32 = 1
-		teMemory           = "100m"
+		teMemory           = "100Mi"
```
Comment (vegichan) [inline test/e2e/nuodb_scaleup_test.go]: same comment as above. Maybe 250MBi if you really need to go that low? You should have 2Gbi available, so you should be able to fit 3 engines with 500MBi each.

File: test/e2e/nuodb_scaleup_test.go [hunk 8]
```diff
@@ -60,17 +66,23 @@ func TestNuodbScale(t *testing.T) {
...
 		Container:         container,
 	}
 
-	exampleNuodb := testutil.NewNuodbCluster(namespace, clusterSpec)
+	exampleNuodbAdmin := testutil.NewNuodbAdmin(namespace, nuodbAdminSpec)
 	testutil.SetupOperator(t,ctx)
+	err = testutil.DeployNuodbAdmin(t, ctx, exampleNuodbAdmin )
+	if err != nil {
+		t.Fatal(err)
+	}
```
Comment (vegichan) [inline test/e2e/nuodb_scaleup_test.go]: assertNilError

File: test/e2e/nuodb_scaleup_test.go [hunk 9]
```diff
@@ -60,17 +66,23 @@ func TestNuodbScale(t *testing.T) {
...
+
+	exampleNuodb := testutil.NewNuodbDatabase(namespace, nuodbSpec)
 	err = testutil.DeployNuodb(t, ctx, exampleNuodb )
 	if err != nil {
 		t.Fatal(err)
 	}
```
Comment (vegichan) [inline test/e2e/nuodb_scaleup_test.go]: assertNilError

File: test/e2e/nuodb_scaleup_test.go [hunk 10]
```diff
@@ -86,12 +98,17 @@ func TestNuodbScale(t *testing.T) {
...
+	err = e2eutil.WaitForDeployment(t, f.KubeClient, namespace, "nuodb-te", 2, testutil.RetryInterval, testutil.Timeout)
 	if err != nil {
 		t.Fatal(err)
 	}
 	t.Log("Te scaled up successfully")
 
+	err = f.Client.Get(goctx.TODO(), types.NamespacedName{Name: "nuodb", Namespace: namespace}, exampleNuodb)
+	if err != nil {
+		t.Fatal(err)
+	}
```
Comment (vegichan) [inline test/e2e/nuodb_scaleup_test.go]: assertNilError

File: test/e2e/util/cluster.go [hunk 11]
```diff
@@ -28,20 +29,51 @@ const (
 	Timeout              = time.Second * 90
 	CleanupRetryInterval = time.Second * 1
 	CleanupTimeout       = time.Second * 15
+	InsightsServerCleanupTimeout = time.Second * 120
 )
 
-func NewNuodbCluster(namespace string, clusterSpec nuodb.NuodbSpec) *nuodb.Nuodb {
+func NewNuodbAdmin(namespace string, adminSpec nuodb.NuodbAdminSpec) *nuodb.NuodbAdmin {
```
Comment (vegichan) [inline test/e2e/util/cluster.go]: why does this return a pointer?

File: test/e2e/util/cluster.go [hunk 12]
```diff
@@ -74,7 +112,25 @@ func SetupOperator(t *testing.T, ctx *framework.TestCtx) {
...
-// DeployCluster creates a custom resource and checks if the
+// DeployNuodbAdmin creates a custom resource and checks if the
+// admin statefulset is deployed successfully.
+func DeployNuodbAdmin(t *testing.T, ctx *framework.TestCtx, nuodbAdmin *nuodb.NuodbAdmin) error {
+	f := framework.Global
+
+	err := f.Client.Create(goctx.TODO(), nuodbAdmin, &framework.CleanupOptions{TestContext: ctx, Timeout: CleanupTimeout, RetryInterval: CleanupRetryInterval})
+	if err != nil {
+		return err
+	}
```
Comment (vegichan) [inline test/e2e/util/cluster.go]: assertNilError

File: test/e2e/util/cluster.go [hunk 13]
```diff
@@ -74,7 +112,25 @@ func SetupOperator(t *testing.T, ctx *framework.TestCtx) {
...
+
+	err = WaitForStatefulSet(t, f.KubeClient, nuodbAdmin.Namespace, "admin", 1, RetryInterval, Timeout*2)
```
Comment (vegichan) [inline test/e2e/util/cluster.go]: why Timeout*2?  Ashutosh had various global timeouts set. If you want 180sec, it should be a separate enum value.

**Comment by vegichan** [comment]
The tests do not follow basic testing hygiene, but I understand that they did not do so from the beginning. We should not chain multiple test cases in one (see scale tests)

File: test/e2e/util/cluster.go (hunk 11 above)
Comment (thebithead) [reply test/e2e/util/cluster.go]: Because the callers and all functions used by the callers are expecting a pointer.

**Pull request merged**